lib/Conversion/TorchToLinalg/Linear.cpp

```diff
@@ -167,11 +171,89 @@ class ConvertAtenMatmulOp : public OpConversionPattern<AtenMatmulOp> {
+    // linalg.batch_matmul op.
+    if (lhsRank == 3 && rhsRank == 3) {
+      Value batchDim = getDimOp(rewriter, loc, lhs, 0);
+      Value lhsDim0 = getDimOp(rewriter, loc, lhs, 1);
+      Value lhsDim1 = getDimOp(rewriter, loc, lhs, 2);
+      Value rhsDim0 = getDimOp(rewriter, loc, rhs, 1);
+      Value rhsDim1 = getDimOp(rewriter, loc, rhs, 2);
+      checkDimEqualHelper(rewriter, loc, lhsDim1, rhsDim0);
+
+      Value zeroTensor = createZeroInitTensor(
+          rewriter, loc, ValueRange{batchDim, lhsDim0, rhsDim1}, elementType);
+      Value matmul =
+          rewriter
+              .create<linalg::BatchMatmulOp>(loc, zeroTensor.getType(),
+                                             ValueRange{lhs, rhs}, zeroTensor)
+              .getResult(0);
+      rewriter.replaceOpWithNewOp<tensor::CastOp>(op, newResultType, matmul);
+      return success();
+    }
+
     // Fourth Case: Batch-Matrix Multiplication.
     // TODO: Broadcasting of batch dimension is remaining.
-    if (lhsRank >= 3 && rhsRank >= 3 && lhsRank == rhsRank) {
+    if (lhsRank > 3 && rhsRank > 3 && lhsRank == rhsRank) {
 
       unsigned batchRank = lhsRank - 2;
+
+      // Incase of static sizes, we can collapse the batchDim and still lower to
+      // linalg.batch_matmul op.
+      if (lhsType.hasStaticShape() && rhsType.hasStaticShape()) {
+
+        ArrayRef<int64_t> lhsShape = lhsType.getShape();
+        ArrayRef<int64_t> rhsShape = rhsType.getShape();
+        ArrayRef<int64_t> resultShape = resultType.getShape();
+
+        SmallVector<int64_t> collapsedResultShape;
+
+        for (int i = 0; i < lhsRank; i++) {
+          if (i < batchRank) {
+            if (collapsedResultShape.empty())
+              collapsedResultShape.push_back(resultShape[i]);
+            else
+              collapsedResultShape[0] *= resultShape[i];
+          } else {
+            collapsedResultShape.push_back(resultShape[i]);
+          }
+        }
+
+        auto collapsedResultType =
+            RankedTensorType::get(collapsedResultShape, elementType);
+
+        // Collapse the batch dimension into one dimension. The resultant rank
+        // will always be 3.
+        SmallVector<ReassociationIndices> reassociation(3);
+
+        for (unsigned i = 0, j = 0; i < lhsRank; i++) {
+          if (i < batchRank) {
+            reassociation[j].push_back(i);
+          } else {
+            j += 1;
+            reassociation[j].push_back(i);
+          }
+        }
+        Value collapsedLhs = rewriter.create<tensor::CollapseShapeOp>(
+            op->getLoc(), lhs, reassociation);
+        Value collapsedRhs = rewriter.create<tensor::CollapseShapeOp>(
+            op->getLoc(), rhs, reassociation);
+        Value initTensor0 = createStaticZeroInitTensor(
+            rewriter, loc, collapsedResultShape, elementType);
+
+        Value batchMatMul =
+            rewriter
+                .create<linalg::BatchMatmulOp>(
+                    loc, collapsedResultType,
```

Comment:
this could be replaced with `initTensor0.getType` and the calculation for `collapsedResultType` could be removed. This would simplify future changes to support dynamic shapes.